src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extensions.cs

```diff
@@ -0,0 +1,539 @@
+
+            foreach (var pair in allBaseExtensions)
+            {
+                NamedTypeSymbol referenceBaseExtension = pair.Key;
+                MultiDictionary<NamedTypeSymbol, NamedTypeSymbol>.ValueSet allBaseExtensionPerCLRSignature = pair.Value;
+
+                foreach (var baseExtension in allBaseExtensionPerCLRSignature)
+                {
+                    baseExtension.CheckAllConstraints(DeclaringCompilation, conversions, location, diagnostics);
+
+                    // PROTOTYPE confirm what we allow in terms of variation between various underlying types
+                    var baseUnderlyingType = baseExtension.ExtensionUnderlyingTypeNoUseSiteDiagnostics;
+                    if (baseUnderlyingType?.Equals(underlyingType, TypeCompareKind.ConsiderEverything) == false)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_UnderlyingTypesMismatch, location, this, underlyingType!, baseUnderlyingType);
+                    }
+
+                    if (!ReferenceEquals(referenceBaseExtension, baseExtension))
+                    {
+                        Debug.Assert(!referenceBaseExtension.Equals(baseExtension, TypeCompareKind.ConsiderEverything));
+                        Debug.Assert(referenceBaseExtension.Equals(baseExtension, TypeCompareKind.CLRSignatureCompareOptions));
+
+                        ReportDuplicate(referenceBaseExtension, baseExtension, location, diagnostics, forBaseExtension: true);
+                    }
+                }
+            }
+        }
+
+        private MultiDictionary<NamedTypeSymbol, NamedTypeSymbol> ExtensionsAndTheirBaseExtensionsNoUseSiteDiagnostics
+        {
+            get
+            {
+                var baseExtensions = this.BaseExtensionsNoUseSiteDiagnostics;
+                var resultBuilder = new MultiDictionary<NamedTypeSymbol, NamedTypeSymbol>(baseExtensions.Length,
+                    SymbolEqualityComparer.CLRSignature, SymbolEqualityComparer.ConsiderEverything);
+
+                foreach (var baseExtension in baseExtensions)
+                {
+                    if (resultBuilder.Add(baseExtension, baseExtension))
+                    {
+                        // PROTOTYPE we need to collect all base extensions from baseExtension too
+                    }
+                }
+
+                return resultBuilder;
+            }
+        }
+
+        internal sealed override TypeSymbol? GetDeclaredExtensionUnderlyingType()
+            => GetDeclaredExtensionInfo().UnderlyingType;
+
+        internal sealed override ImmutableArray<NamedTypeSymbol> GetDeclaredBaseExtensions()
+            => GetDeclaredExtensionInfo().BaseExtensions;
+
+        protected sealed override TypeSymbol? ExtensionUnderlyingTypeNoUseSiteDiagnosticsCore
+        {
+            get
+            {
+                if (ReferenceEquals(_lazyExtensionUnderlyingType, ErrorTypeSymbol.UnknownResultType))
```

Comment:
:memo: I want to point out a choice that I made here and get your opinion.
If you look at `BaseTypeNoUseSiteDiagnostics`, which has similar logic, it pulls on the base type of the containing type. 
I couldn't think of why that would be useful for the underlying type, so didn't implement such logic here.

```
        internal sealed override NamedTypeSymbol BaseTypeNoUseSiteDiagnostics
        {
            get
            {
                if (ReferenceEquals(_lazyBaseType, ErrorTypeSymbol.UnknownResultType))
                {
                    // force resolution of bases in containing type
                    // to make base resolution errors more deterministic
                    if ((object)ContainingType != null)
                    {
                        var tmp = ContainingType.BaseTypeNoUseSiteDiagnostics;
                    }
...
```